pennylane/devices/device_api.py

```diff
@@ -0,0 +1,447 @@
+            error and a postprocessing function to be called after execution.
+
+        Raises:
+            Exception: An exception is raised if the input cannot be converted into a form supported by the device.
+
+        This function is tracked by machine learning interfaces and should be fully differentiable.
+
+        Preprocessing steps may include:
+        - expansion to :class:`~.Operator`'s and :class:`~.MeasurementProcess` objects supported by the device.
+        - splitting a quantum script with the measurement of non-commuting observables or Hamiltonians into multiple executions
+        - splitting quantum scripts with batched parameters into multiple executions
+
+        This step may also validate parameters of the ``execution_config``. For example, it could raise an error if the
+        device doesn't support finite shots but the ``execution_config`` requests it.
+
+        Processing steps can depend on the requested gradient method.
+
+        """
+
+        def blank_postprocessing_fn(res):
+            """Identity postprocessing function created in QuantumDevice preprocessing.
+
+            Args:
+                res (tensor-like): A result object
+
+            Returns:
+                tensor-like: The function input.
+
+            """
+            return res
+
+        qscript_batch = [qscript] if isinstance(qscript, QuantumScript) else qscript
+        return qscript_batch, blank_postprocessing_fn
+
+    @abc.abstractmethod
+    def execute(self, qscripts: QScript_or_QScriptBatch, execution_config=None):
+        """Execute a Quantum Script and turn it into a result.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            A numeric result of the computation.
+
+        **Execution Config:**
+
+        The execution config can be:
+
+        * ``None``: device defaults should be used
+
+        * A single ``ExecutionConfig`` datastructure: All scripts in the batch use the same configuration
+
+        * An Iterable of ``ExecutionConfig``'s of the same length as the the qscripts input: Each quantum script
+          has its own set of configuration parameters. For example, every script can be executed with a different
+          number of shots.
+
+        **Interface parameters:**
+
+        Note that the parameters contained within the quantum script may contain interface-specific data types, such as
+        ``torch.tensor`` or ``tensorflow.Variable``. If the device does not wish to handle interface-specific parameters, they
+        can implement an optional "internal preprocessing" step that converts all parameters to vanilla numpy. A convenience
+        transform implementing this will be provided.
```

Comment:
@albi3ro the script containing tensor parameters can _only_ occur if the diff method is `backprop`, right? In which case, the device should somehow specify 'I support backprop?'

Otherwise, PL would pick `diff_method=some_grad_transform`, and the script parameters should always be NumPy by this point.

So: is there ever a case where the device gets a script with tensors, but doesn't natively support those tensors?

Reply:
For this interface, I want to push any decision that depends on the device inside one of these main methods.

We have two options:

1) Add flags for every capability or quirk where devices could differ.  Then modify code elsewhere in pennylane based on those flags. This is what the device currently does. It makes it hard for devices to make a different decision, and they would have to add a flag and then change the pennylane core source code.

2) Make the device itself handle any decision that could vary between devices. Then provide optional tools to handle the standard cases. This is what we should do instead.

I'd rather say "the parameters could belong to any interface" and provide a `convert_to_numpy_parameters` transform to use.  Otherwise, we have to increase the complexity of the interface by adding a flag that says "convert to numpy for me please". 

And I don't think tensor parameters should only occur in backpropagation. What if someone wants to jit parameter shift and adjoint differentiation? ML framework parameters are useful for more than just differentiation.

Reply:
Execution that happens in the interfaces always unwrap the tapes before executing, for jit and parameter shift we never deal we ML tensors in the execution. The only time we do not unwrap tapes is for backpropagation.